api-select.go

```diff
@@ -0,0 +1,589 @@
+	})
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, httpRespToErrorResponse(resp, bucketName, "")
+	}
+
+	// Process needs to return a eventStream to the client
+	// Wrap this in a Go routine so as to allow the streaming
+	streamer := &Events{
+		Response: make(chan interface{}),
+	}
+	go func() {
+		processBinary(resp.Body, streamer)
+	}()
+	return streamer, nil
+}
+
+// processBinary is the main function that decodes the large byte array into
+// several events that are sent through the eventstream.
+func processBinary(message io.Reader, streamer *Events) {
+	var partialRecordVal []byte
+	lenOfTrim := 0
+	for {
+		var prelude preludeInfo
+		var headers *headerInfo
+		// Create CRC code
+		crc := crc32.New(crc32.IEEETable)
+		crcReader := io.TeeReader(message, crc)
+		var err error
+		var payload []byte
+		// Extract the prelude(12 bytes) into a struct to extract relevant
+		// information.
+		prelude, err = processPrelude(crcReader, crc)
+		if err != nil {
+			if streamer.err == nil {
+				streamer.err = err
+			}
+			continue
+		}
+		// Extract the headers(variable bytes) into a struct to extract relevant
+		// information
+		if prelude.headerLen > 0 {
+			headers, err = extractHeader(io.LimitReader(crcReader, int64(prelude.headerLen)))
+			if err != nil {
+				if streamer.err == nil {
+					streamer.err = err
+				}
+			}
+		}
+		// Calculate the length of the payload so that the appropriate amount of
+		// bytes can be read
+		payloadLen := calcPayloadLen(prelude)
+		// Checks that there is a payload and if there is, then extracts it.
+		if payloadLen > 0 {
+			payload, err = extractPayload(io.LimitReader(crcReader, int64(payloadLen)), payloadLen)
+			// Special case for dealing with messages that are partial.
+			if isPartial(payload) {
+				payload, partialRecordVal, lenOfTrim = processPartial(payload)
+				processEvent(payload, headers, streamer)
+				// return the complete part of the message so far to the client
+				// for loop in case there are multiple sequential partial rows
+				for {
+					// Extracts the partial payload.
+					myPayload, err := extractPayload(io.LimitReader(crcReader, int64(lenOfTrim)), lenOfTrim)
+					if err != nil {
+						if streamer.err == nil {
+							streamer.err = err
+						}
+					}
+					// knits together the payload with the previous partial one.
+					myPayload = append(partialRecordVal, myPayload...)
+					if isPartial(myPayload) {
+						// if its another partial payload we need to write current complete
+						// part and set part that needs to be stitched
+						var completePayload []byte
+						completePayload, partialRecordVal, lenOfTrim = processPartial(myPayload)
+						processEvent(completePayload, headers, streamer)
+						continue
+					}
+					// if there is no longer a partial porition we can return our payload
+					// to the client and break
+					if len(myPayload) > 0 {
+						processEvent(myPayload, headers, streamer)
+					}
```

Comment:
Most calls to `processEvent` are ignoring the returned value. This does not look correct when considering error handling. Also on error, it is required to close the Events.Response channel, so that the for loop that is consuming the events quits - actually this does not seem to be happening in the success case, when all results have been returned and the program should exit.

See https://play.golang.org/p/wIARKfwXJPv to understand the problem